File: HW/HW07(Java)/src/AbstractBoard.java
import java.io.FileNotFoundException;
import java.io.IOException;

public abstract class AbstractBoard implements Cloneable{
	protected int height;
	protected int width;
	protected char lastMove;
	protected int numberMoves;
	protected static int numberBoards = 0;

	public AbstractBoard(int size1,int size2){
		height = size1;
		width = size2;
		lastMove = 's';
		numberMoves = 0;	}
	public AbstractBoard(){
		height = 0;
		width = 0;
		lastMove = 's';
		numberMoves = 0;
	}
	public int getHeight(){
		return height;
	}
	public int getWidth(){
		return width;
	}
	public char getLastMove(){
		return lastMove;
	}
	public int getNumberMoves(){
		return numberMoves;
	}
	public int getNumberBoards(){
		return numberBoards;
	}
	public void setSize(int size1,int size2){
		height = size1;
		width = size2;
	}
	public String toString(){
		return String.format("Height: %d",height," Width: %d\n",width,
			"Number Of Boards: %d\n",numberBoards,
			"Number Of Moves: %d\n",numberMoves,
			"Last Moves: %c\n",lastMove);
	}
	public int invCount(int board[],int size){
		int count = 0;
        for (int i=0;i<size-1;i++)
            for (int j=i+1;j<size;j++)
                if(board[i]>board[j])
                    count++;
        return count;
	}
	public boolean solved(int board[],int size){
		int count = invCount(board,size);
        return (count % 2 == 0);
	}
	public Object clone()throws CloneNotSupportedException{  
		return super.clone();  
	}
	public abstract void readFromFile(String answer)throws FileNotFoundException;
	public abstract void writeToFile(String answer)throws IOException;
	public abstract void reset();
	public abstract void move(char move);
	public abstract boolean isSolved();
	public abstract int cell(int x,int y);
	public abstract boolean equals(AbstractBoard obje);
}

File: HW/HW05/AbstractBoard.cpp
#include "AbstractBoard.h"
#include <iostream>
#include <fstream>
#include <string>
#include <cstdlib>
#include <ctime>
#include <cmath>
#include <vector>
using namespace std;

namespace Board{
    int AbstractBoard::numberOfBoard = 0;
    AbstractBoard::AbstractBoard():height(0),width(0),numberOfMove(0),lastMove('s'){
    }
    AbstractBoard::AbstractBoard(int size1,int size2):height(size1),width(size2),numberOfMove(0),lastMove('s'){
    }
    AbstractBoard::AbstractBoard(const AbstractBoard &obje):height(obje.getHeight()),width(obje.getWidth()),numberOfMove(obje.getNumberOfMoves()),lastMove(getLastMove()){
        numberOfBoard = obje.getnumberOfBoards();
    }
    int AbstractBoard::getHeight()const{
        return height;
    }
    int AbstractBoard::getWidth()const{
        return width;
    }
    int AbstractBoard::getNumberOfMoves()const{
        return numberOfMove;
    }
    int AbstractBoard::getnumberOfBoards()const{
        return numberOfBoard;
    }
    char AbstractBoard::getLastMove()const{
        return lastMove;
    }
    void AbstractBoard::setSize(int size1, int size2){
        height = size1;
        width = size2;
    }
    int AbstractBoard::invCount(int board[], int size){
        int count = 0;
        for (int i=0;i<size-1;i++)
            for (int j=i+1;j<size;j++)
                if(board[i]>board[j])
                    count++;
        return count;
    }
    bool AbstractBoard::solved(int board[], int size){
        int count = invCount(board,size);
        return (count % 2 == 0);
    }
    AbstractBoard& AbstractBoard::operator=(const AbstractBoard &obje){
        height = obje.getHeight();
        width = obje.getWidth();
        numberOfMove = obje.getNumberOfMoves();
        lastMove = obje.getLastMove();
        numberOfBoard = obje.getnumberOfBoards();
        return *this;
    }
    AbstractBoard::~AbstractBoard(){
        numberOfBoard--;
    }
    
    bool movesForSolution(vector<AbstractBoard *> boards){
        int numberBoard = boards[0]->getnumberOfBoards();
        bool control = true;
        AbstractBoard *temp = boards[numberBoard];
        temp->reset();
        for(int i=numberBoard;i>0;i--){
            if(temp == boards[i]){
                char move = '\0';
                if(boards[i-1]->getLastMove() == 'd')
                    move = 'u';
                if(boards[i-1]->getLastMove() == 'u')
                    move = 'd';
                if(boards[i-1]->getLastMove() == 'r')
                    move = 'l';
                if(boards[i-1]->getLastMove() == 'l')
                    move = 'r';
                if(boards[i-1]->getLastMove() == 's'){
                    control = false;
                    break;
                }
                temp->move(move);
            }
            else{
                control = false;
                break;
            }
        }
        return control;
    }
}




